### PR TITLE
feat(tree2): allow handles in ops and add GC tests

### PR DESCRIPTION
**Pull request by connorskees**
Uses `IFluidSerializer.encode` and `IFluidSerializer.decode` to bind the handles without serializing the contents, avoiding the cost of double stringification and avoiding a crash.

Also adds GC tests, which verify encoding+decoding of handles works successfully.

### CODE REVIEW

**Comment by connorskees** [comment]
This PR is not ready for review, as it is missing tests and code for the decode step.

File: packages/dds/sequence/src/test/sharedString.spec.ts [hunk 1]
```diff
@@ -709,6 +709,43 @@ describe("SharedString", () => {
 			// Verify that the changes were correctly received by the second SharedString
 			assert.equal(sharedString2.getText(), "hello friend");
 		});
+		
+		it("insert in middle of multibyte character", async () => {
```
Comment (connorskees) [inline packages/dds/sequence/src/test/sharedString.spec.ts:713]: this snuck in here on accident, but i don't think it's harmful to include, or at least harmful enough to warrant splitting out to a separate PR.

File: experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts [hunk 2]
```diff
@@ -260,15 +260,16 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
 			commit,
 			sessionId: this.editManager.localSessionId,
 		});
-		this.submitLocalMessage(message);
+		this.submitLocalMessage(this.serializer.encode(message, this.handle));
```
Comment (tyler-cai-microsoft) [inline experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts:263]: This is interesting, I wonder what GC implications this has. @markfields @agarwal-navin, this is how SharedTree is doing their handle work

**Review by tyler-cai-microsoft — APPROVED** [review]
:shipit: LGTM

File: experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts [hunk 3]
```diff
@@ -96,6 +96,21 @@ describe("SharedTree", () => {
...
+		assert(provider.trees[1].isAttached());
+
+		const field = provider.trees[0].editor.optionalField({
+			parent: undefined,
+			field: rootFieldKey,
+		});
+		field.set(
+			singleTextCursor({ type: leaf.handle.name, value: provider.trees[0].handle }),
+			true,
+		);
```
Comment (tyler-cai-microsoft) [inline experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Can you test blob handles? I think they are quite relevant for `SharedTree`. They are more likely to be used than the other handle types

Comment (connorskees) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Blob handle?

Comment (connorskees) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Or, rather, how does the handling of a blob handle differ from a DDS handle?

Comment (tyler-cai-microsoft) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Typing, the difference is typing. That's it

Comment (tyler-cai-microsoft) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Blob handles are of type `IFluidHandle<ArrayBufferLike>` which fails the type check of `IFluidHandle<FluidObject & IFluidLoadable>` as `ArrayBufferLike` does not have the property `IFluidLoadable`

Comment (connorskees) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: Is there an existing example of tests for blob handles? I only see e2e+gc tests that don't really test specific DDSes

Comment (tyler-cai-microsoft) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: The only reason I recommend this testing for `SharedTree` is that likely the type of `IFluidHandle` stored in `SharedTree` will be `BlobHandles`. Other than that, I'm not particularly biased. Feel free to do this in a separate PR, I don't want to block this one.

File: experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts [hunk 4]
```diff
@@ -132,10 +132,15 @@ export type FluidSerializableReadOnly =
 	| { readonly [P in string]?: FluidSerializableReadOnly };
 
 // TODO: replace test in FluidSerializer.encodeValue with this.
-export function isFluidHandle(value: undefined | FluidSerializableReadOnly): value is IFluidHandle {
+export function isFluidHandle(value: unknown): value is IFluidHandle {
 	if (typeof value !== "object" || value === null) {
 		return false;
 	}
+
+	if (!("IFluidHandle" in value)) {
```
Comment (Abe27342) [inline experimental/dds/tree2/src/feature-libraries/contextuallyTyped.ts:140]: the check on L147 seems stricter. Why do we need this as well?

File: experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts [hunk 5]
```diff
@@ -260,15 +260,16 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
...
 	}
 
 	protected processCore(
 		message: ISequencedDocumentMessage,
 		local: boolean,
 		localOpMetadata: unknown,
 	) {
-		const { commit, sessionId } = this.messageCodec.decode(message.contents);
+		const contents = this.serializer.decode(message.contents);
```
Comment (Abe27342) [inline experimental/dds/tree2/src/shared-tree-core/sharedTreeCore.ts:271]:
```suggestion
		const contents: unknown = this.serializer.decode(message.contents);
```

the typing on IFluidSerializer is extremely generous and it's easy to make mistakes. Logic here is sound, but generally try to declare an as-specific-type-as-possible when using return value of IFluidSerializer methods.

File: experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts (hunk 3 above)
Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/sharedTree.spec.ts:111]: I don't think the runtime behavior is meaningfully different from the DDS perspective (it's not like any fluid handles we're testing today will be deref'd by SharedTree itself, just application code), but seems fine to add a separate test afterward for this. It's just a bit annoying as unit tests don't support the blob APIs so the set of things you need to mock is relatively large, at which point the test has a bit less value.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**